amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseCoordinator.java

```diff
@@ -329,8 +329,9 @@ public void stop() {
 
     @Override
     public void stopLeaseTaker() {
-        takerFuture.cancel(false);
-
+        if (takerFuture != null) {
+            takerFuture.cancel(false);
```

Comment:
In handling this null case is there any chance that takerFuture is assigned to Null. Do we need to mark this somewhere?

Reply:
> is there any chance that takerFuture is assigned to Null.

Yes, `takerFuture=null` at declaration and remains so until reassigned @ https://github.com/awslabs/amazon-kinesis-client/blob/master/amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseCoordinator.java#L236-L239  This is currently the only assignment, so it will not revert to `null`.

> Do we need to mark this somewhere?

Would you kindly clarify what is being asked?

Reply:
Trying to understand the RCA for the bugs, first one says if scheduler is shutdown too fast after started, I dont think thats possible because start and shutdown are synchronized on the scheduler object. But its possible that if shutdown happened first before start could even run, but then we dont have the right behavior either, ideally if shutdown ran first and then start runs, we should disallow the start, so im thinking maybe we want to keep track of a flag for started and shutdown and only do shutdown processing if it was started and dont allow a start if it was already shutdown?

I dint quite understand the second linked issue from a quick look.

Reply:
> [aravinda] I think maybe thats what you called out in your description but a bias-for-action fix is to address NPE?

Yeah, replication paths weren't provided in the issues but the stacktraces consistently implicated this line.

I agree that we _should_ "disallow the start", and other more defensive/correct behaviors. However, time is a luxury and this bias-for-action PR is better than letting this linger another ~2.5yrs.

Reply:
I agree with bias-for-action if it fixes the issues referenced, but maybe they dont see the NPE with this fix, but still their application wont behave as they want which I assume is, cancel the start and shutdown gracefully? So is there value in doing this, i.e. if we have correct guarantees about only shutdown if started and only start if not already shutdown then this null check wont even be needed? Just a thought, please let me know what you think.

Reply:
I was just asking if for some reason the function returned null do we need to log this for future reference?